### DISCUSSION TITLE
Share database between multiple kamal applications though accessories or another method ?

### DISCUSSION THREAD

**Discussion by rogerheykoop** [basecamp/kamal#1085] [Q&A]
I am de-clouding a bunch of existing Rails applications. These used to run on AWS ECS, and they used to connect to an RDS database server. 

For that goal I've got a new dedicated server. 
Suppose I have two separate Rails applications A and B, each will be deployed to that same new server using Kamal 2. 

If I add a MySQL service as an accessory to A, and I name its service 'shared-mysql' for instance, then I'm not able to add the same containerized MySQL service to B as 'shared-mysql', or am I mistaken ?

What would be the best way to approach this in a Kamal 2 based setup ? 
Add MySQL as a native service on the dedicated server and connect to it from the containers ? Or is there another way ?

**Comment by sasharevzin** [comment]
My understanding is that you’ll need to have the same network name to be able to connect from both apps

**Reply by rogerheykoop** [reply]
Kamal creates its own network. 
I've figured out how to make this work. 

What I did was this in App A:

 db:
    image: mysql:8.0
    service: shared-mysql
    port: '127.0.0.1:3306:3306'
    host: ******************
    env:
      clear:
        MYSQL_ROOT_HOST: '%'
      secret:
        - MYSQL_ROOT_PASSWORD
    files:
      - db/development.sql:/docker-entrypoint-initdb.d/setup.sql
    directories:
      - data/mysql:/var/lib/mysql

      
 And then this in App B:
 
   db:
    service: shared-mysql
    host: ******************************
    env:
      clear:
        MYSQL_ROOT_HOST: '%'
      secret:
        - MYSQL_ROOT_PASSWORD
    files:
      - db/development.sql:/docker-entrypoint-initdb.d/setup.sql
    directories:
      - data/mysql:/var/lib/mysql
      
So basically you create an accessory with a service name in A, then reference that service by not using an image in B.

**Reply by hboon** [reply]
@rogerheykoop I tried this with Postgres, and it doesn't work for me. I'm using Kamal 2.2.2. Which version did you use it with?

It either goes:

```
docker stderr: "docker run" requires at least 1 argument.
```

or:

```
docker stderr: docker: Error response from daemon: Conflict. The container name "/shared-database" is already in use by container "f71477e85cd7f06e1d97681b80ae7a1a4e5b794e12f36380a1415162c8805c65". You have to remove (or rename) that container to be able to reuse that name.
```

Configuration:

```
accessories:
  db:
    image: postgres:14
    service: shared-database
    host: <ip>
    port: 5432
    env:
      clear:
        POSTGRES_HOST_AUTH_METHOD: password
      secret:
        - POSTGRES_USER
        - POSTGRES_DB # The database will be created automatically on first boot.
        - POSTGRES_PASSWORD
    volumes:
      - "/psql/14/main/var/lib/postgresql/data"
```

and

```
accessories:
  db:
    #image: postgres:14
    service: shared-database
    host: <ip>
    port: 5432
    env:
      clear:
        POSTGRES_HOST_AUTH_METHOD: password
      secret:
        - POSTGRES_USER
        - POSTGRES_DB # The database will be created automatically on first boot.
        - POSTGRES_PASSWORD
    volumes:
      - "/psql/14/main/var/lib/postgresql/data"
```

**Reply by cesarho** [reply]
@hboon I'm interested to know how you resolved this issue.

**Reply by nickhammond** [reply]
@rogerheykoop I haven't tried this but I would think you'd just want one app to be the manager of the db accessory due to the naming conflicts mentioned in the comments.

So app A would define and control the db accessory. App B wouldn't know about it in the kamal sense, but it's going to be on the same `kamal` network, so your MySQL host in your app containers would be `#{service}-db`.

**Reply by achilles-tee** [reply]
```
docker stderr: "docker run" requires at least 1 argument.
docker stderr: docker: Error response from daemon: Conflict. The container name...
```
I encountered the above errors as well. I found that you don't necessary to run kamal setup for app B since the db is already setup in app A. after I performed kamal deploy, the app managed to access the postgres DB. 
The accessories setup just requires the service, host, port, env

**Comment by kevinluo201** [comment]
I solved this issue by creating another Kamal project solely for the database. Given that I have two apps, `app_a` and `app_b`, and I want both of them to use Postgres 17 as their database. I'll create a new folder, `shared_postgres17`, and enter `kamal init` to initiate the kamal project. I then only configure the accessories in its `deploy.yml`:
```yaml
service: shared_postgres17

### START OF --- required fields, I don't think they're needed, they're just required by kamal
image: postgres:17.5
registry:
  username: my_username
  password:
    - KAMAL_REGISTRY_PASSWORD
builder:
  arch: amd64
### END OF --- required fields

ssh:
  keys_only: true
  keys: [ "~/.ssh/my_key" ]
  config: true

accessories:
  db:
    image: postgres:17.5
    host: you_ip_address
    port: 5432
    env:
      secret:
        - POSTGRES_PASSWORD
    directories:
      - data:/var/lib/postgresql/data

```

I can execute the following command to deploy the database. Of course, Docker needs to be installed on the server beforehand.
```bash
kamal accessories boot db
```

In the `deploy.yml` of `app_a` and `app_b`, I'll let them connect to `shared_postgres17-db` because it will be the domain name of the shared database.

```yaml
# assume it's the content of app_a's deploy.yml
env:
  clear:
    POSTGRES_HOST: shared_postgres17-db
    POSTGRES_PORT: 5432
    POSTGRES_DB: app_a_db
    POSTGRES_USER: postgres
  secret:
    - RAILS_MASTER_KEY
    - POSTGRES_PASSWORD
```
Then they can use the same database on the same server. The drawback of this setting is that there is no accessories section in their `deploy.yml` anymore. The deployers have to know that the database is deployed separately. (Or, always boot the database in hooks, like `pre-app-boot`)
